homeassistant/components/comfoair/sensor.py

```diff
@@ -0,0 +1,157 @@
+
+ATTR_COMFORT_TEMPERATURE = "comfort_temperature"
+ATTR_OUTSIDE_TEMPERATURE = "outside_temperature"
+ATTR_SUPPLY_TEMPERATURE = "supply_temperature"
+ATTR_RETURN_TEMPERATURE = "return_temperature"
+ATTR_EXHAUST_TEMPERATURE = "exhaust_temperature"
+ATTR_AIR_FLOW_SUPPLY = "air_flow_supply"
+ATTR_AIR_FLOW_EXHAUST = "air_flow_exhaust"
+ATTR_FAN_SPEED_MODE = "speed_mode"
+
+SENSOR_TYPES = {
+    ATTR_COMFORT_TEMPERATURE: [
+        "Comfort Temperature",
+        TEMP_CELSIUS,
+        "mdi:thermometer",
+        0xD2,
+        0 * 8,
+        8,
+    ],
+    ATTR_OUTSIDE_TEMPERATURE: [
+        "Outside Temperature",
+        TEMP_CELSIUS,
+        "mdi:thermometer",
+        0xD2,
+        1 * 8,
+        8,
+    ],
+    ATTR_SUPPLY_TEMPERATURE: [
+        "Supply Temperature",
+        TEMP_CELSIUS,
+        "mdi:thermometer",
+        0xD2,
+        2 * 8,
+        8,
+    ],
+    ATTR_RETURN_TEMPERATURE: [
+        "Return Temperature",
+        TEMP_CELSIUS,
+        "mdi:thermometer",
+        0xD2,
+        3 * 8,
+        8,
+    ],
+    ATTR_EXHAUST_TEMPERATURE: [
+        "Exhaust Temperature",
+        TEMP_CELSIUS,
+        "mdi:thermometer",
+        0xD2,
+        4 * 8,
+        8,
+    ],
+    ATTR_AIR_FLOW_EXHAUST: ["Exhaust airflow", "%", "mdi:fan", 0xCE, 6 * 8, 8],
+    ATTR_AIR_FLOW_SUPPLY: ["Supply airflow", "%", "mdi:fan", 0xCE, 7 * 8, 8],
+    ATTR_FAN_SPEED_MODE: ["Speed mode", "", "mdi:fan", 0xCE, 8 * 8, 8],
+}
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the ComfoAir sensor platform."""
+    unit = hass.data[DOMAIN]
+
+    sensors = []
+    for resource in config.get(CONF_RESOURCES, []):
+        sensor_type = resource.lower()
+        _LOGGER.debug("Sensor type: %s", sensor_type)
+
+        if sensor_type not in SENSOR_TYPES:
+            _LOGGER.warning("Sensor type: %s is not a valid sensor", sensor_type)
+            continue
+
+        sensors.append(
+            ComfoAirSensor(
+                hass,
+                name="%s %s" % (unit.name, SENSOR_TYPES[sensor_type][0]),
+                ca=unit,
+                sensor_type=sensor_type,
+            )
+        )
+
+    async_add_entities(sensors, True)
+
+
+class ComfoAirSensor(Entity):
+    """Representation of a ComfoAir sensor."""
+
+    def __init__(self, hass, name, ca: ComfoAirModule, sensor_type) -> None:
+        """Initialize the ComfoAir sensor."""
+        self._ca = ca
+        self._sensor_type = sensor_type
+        self._unit = SENSOR_TYPES[self._sensor_type][1]
+        self._icon = SENSOR_TYPES[self._sensor_type][2]
+        self._sensor_id = SENSOR_TYPES[self._sensor_type][3]
+        self._offset = SENSOR_TYPES[self._sensor_type][4]
+        self._size = SENSOR_TYPES[self._sensor_type][5]
+        self._name = name
+        self._data = None
+
+        def _update_state(data):
+            bits = BitArray(data)
+            value = bits[self._offset : self._offset + self._size].uint
```

Comment:
We shouldn't need to handle bits and offsets in home assistant. Please move that to the library.